sdk/servicebus/service-bus/test/perf/rhea-promise/receive.ts

```diff
@@ -0,0 +1,157 @@
+/*
+# Overview
+Measures the maximum throughput of `receiver.receive()` in package `rhea-promise`.
+
+# Instructions
+1. Create a Service Bus namespace with `Tier=Premium` and `Messaging Units=4`.  It is recommended to use the largest possible namespace to allow maximum client throughput.
+2. Create a queue inside the namespace.
+3. Set env vars `SERVICE_BUS_CONNECTION_STRING` and `SERVICE_BUS_QUEUE_NAME`.
+4. This test presumes that there are messages in the queue.
+5. `ts-node receive.ts [maxConcurrentCalls] [totalMessages]`
+6. Example: `ts-node receive.ts 1000 1000000`
+ */
+
+import { Connection, ConnectionOptions, ReceiverEvents } from "rhea-promise";
+import delay from "delay";
+import moment from "moment";
+
+const _start = moment();
+
+let _messages = 0;
+
+async function main(): Promise<void> {
+  // Endpoint=sb://<your-namespace>.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=<shared-access-key>
+  const connectionString = process.env.SERVICE_BUS_CONNECTION_STRING as string;
+  const entityPath = process.env.SERVICE_BUS_QUEUE_NAME as string;
+  const allowUnauthorized = process.env.SERVICE_BUS_ALLOW_UNAUTHORIZED ? true : false;
+
+  // <your-namespace>.servicebus.windows.net
+  const host = connectionString.match(/\/\/([^\/]*)\//)![1];
+
+  // SharedAccessKeyName (usually  "RootManageSharedAccessKey")
+  const username = connectionString.match(/SharedAccessKeyName=(.*);/)![1];
+
+  // SharedAccessKey
+  const password = connectionString.match(/SharedAccessKey=(.*)/)![1];
+
+  const maxConcurrentCalls = process.argv.length > 2 ? parseInt(process.argv[2]) : 10;
+  const messages = process.argv.length > 3 ? parseInt(process.argv[3]) : 100;
+  log(`Maximum Concurrent Calls: ${maxConcurrentCalls}`);
+  log(`Total messages: ${messages}`);
+
+  const writeResultsPromise = WriteResults(messages);
+
+  await RunTest(
+    host,
+    username,
+    password,
+    allowUnauthorized,
+    entityPath,
+    maxConcurrentCalls,
+    messages
+  );
+
+  await writeResultsPromise;
+}
+
+async function RunTest(
+  host: string,
+  username: string,
+  password: string,
+  allowUnauthorized: boolean,
+  entityPath: string,
+  maxConcurrentCalls: number,
+  messages: number
+): Promise<void> {
+  const port = 5671;
+
+  const connection = new Connection({
+    transport: "tls",
+    host: host,
+    hostname: host,
+    username: username,
+    password: password,
+    port: port,
+    reconnect: false,
+    rejectUnauthorized: !allowUnauthorized
+  } as ConnectionOptions);
+  await connection.open();
+
+  const receiver = await connection.createReceiver({
+    name: "receiver-1",
+    source: {
+      address: entityPath
+    },
+    credit_window: 0
```

Comment:
Is there a reason we are manually managing credits instead of using the default prefetch window? From the `rhea` docs:

> add_credit(n)
> 
> By default, receivers have a prefetch window that is moved automatically by the library. However if desired the application can set the prefecth to zero and manage credit itself. Each invocation of add_credit() method issues credit for a further 'n' messages to be sent by the peer over this receiving link. [Note: flow()is an alias for add_credit()]
> 
> https://github.com/amqp/rhea/tree/master#add_creditn

Reply:
I don't fully understand how the "credit" concept works for send and receive, but I get the impression it's important for performance.  We may want to even rethink how the `send` tests work.  Perhaps we shouldn't be taking a fixed number for `maxInflight`, and should instead be relying on the "credits" to notify the sender when to stop/resume sending.

Reply:
In the rhea send case, we might not need a fixed `maxInflight`, assuming `sender.sendable()` flips to `false` when there's an appropriate number of sends inflight.  However, in the service-bus send case (where send() returns a Promise), we do need some way to prevent creating too many inflight messages at once.  There needs to be some mechanism to apply backpressure to the sender.  Maybe this is a concept we need to expose in the service-bus API?

Reply:
I will run a quick test of rhea send and receive with and without explicitly managing max inflight to see how they compare.

Reply:
I preferred manual credit management as that is what we do in the service bus library.

I don't foresee any big perf difference between setting `credit_window: maxConcurrentCalls` vs the current way where
- `credit_window` is set to 0
- Add `maxConcurrentCalls` credits
- Add 1 credit every time we receive a message i.e when in the event listener for the `message` event

If we let rhea do the credit management, it does step 2 and then adds a credit each time it fires the `message` event

Reply:
There are two ways to think about a lower-level test (`rhea-promise`) compared to a higher-level test (`service-bus`).

One is to use each library how we recommend customers use it, and then compare the perf.  The goal is to measure the real-world performance difference between the libraries when used in the default/recommended way.

Another is to use the lower-level library the same way it's used by the higher-level library, even if this is different from the default/recommended way we expect customers to use it.  The goal is to measure the overhead added by the higher-level library, holding the lower-level library as constant as possible.

Both tests may be useful but they have different goals.  In this case you may be correct that the lower-level library will have the same perf either way, but ideally we should measure to make sure.

In general, I believe the first approach is a higher priority, since its more meaningful to customers.  The second approach is more useful for isolating and investigating performance issues.

So my recommendation would be for the `rhea-promise` test to use whatever code we expect most customers would use, which I believe is to allow rhea to manage the credits.

Reply:
Another reason it's important to use each library the way we recommend it to customers, is it may expose design issues.  For example, if it's possible to get high performance with `rhea-promise` without manually managing credits, but you do need to manually manage credits when using `service-bus`, this may be a design flaw in the `service-bus` layer we should fix.

Reply:
>In the rhea send case, we might not need a fixed maxInflight, assuming sender.sendable() flips to false when there's an appropriate number of sends inflight. However, in the service-bus send case (where send() returns a Promise), we do need some way to prevent creating too many inflight messages at once. There needs to be some mechanism to apply backpressure to the sender. Maybe this is a concept we need to expose in the service-bus API?

We do this check `sender.sendable()` in the service-bus sdk.
In rhea's send test, `sender.sendable()` guards the test if we don't use `maxInFlight`. (which is not the case with service-bus send test)
Maybe we can consider exposing this in the service-bus API.